### PR TITLE
fully refactored library

### PR DESCRIPTION
**Pull request by AmsTaFFix**
- remove usage of chunks
- mark deprecated all old classes
- use PHP7.1 and strict types in every new class
- changed architecture
- create tests with 100% on part of classes

### CODE REVIEW

File: src/Enum/SubscribersResponse/Type.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php
+declare(strict_types=1);
+
+namespace Citilink\ExpertSenderApi\Enum\SubscribersResponse;
+
+use MyCLabs\Enum\Enum;
+
+/**
+ * Тип свойства подписчика
```
Comment (liderman) [inline src/Enum/SubscribersResponse/Type.php]: Please correct the documentation language in English.

File: src/HttpTransport.php [hunk 2]
```diff
@@ -19,11 +24,18 @@ public function getContext()
...
+     * @param null $content
+     *
+     * @return ResponseInterface
+     */
     public function query($url, $method = 'GET', $content = null)
     {
         $http = [
             'method' => $method,
-            'header' => "Content-Type: text/xml\r\n",
+            'header' => "Content-DataType: text/xml\r\n",
```
Comment (liderman) [inline src/HttpTransport.php]: There is no mistake?

File: src/Model/SubscribersRequest/Value.php [hunk 3]
```diff
@@ -0,0 +1,126 @@
+<?php
+declare(strict_types=1);
+
+namespace Citilink\ExpertSenderApi\Model\SubscribersRequest;
+
+use Citilink\ExpertSenderApi\Enum\DataType;
+
+/**
+ * Value of property
+ */
```
Comment (liderman) [inline src/Model/SubscribersRequest/Value.php]: Doc-block `@author`?

File: src/Model/SubscribersResponse/Value.php [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+     *
+     * @param DataType $expectedType Expected type
+     *
+     * @return ExpertSenderApiException Exception
+     */
+    private function createException(DataType $expectedType): ExpertSenderApiException
+    {
+        return new ExpertSenderApiException(
+            sprintf(
+                'Попытка получить данные неверное типа из свойства подписчика. Ожидаемый тип: %s, Текущий тип %s',
```
Comment (liderman) [inline src/Model/SubscribersResponse/Value.php]: Please use English

File: src/Model/TransactionalRequest/Receiver.php [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    private $emailMd5;
+
+    /**
+     * @var int|null List ID
+     */
+    private $listId;
+
+    /**
+     * Constructor
```
Comment (liderman) [inline src/Model/TransactionalRequest/Receiver.php]: Need more detailed description

File: src/Model/TransactionalRequest/Receiver.php [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @return static Receiver of message
+     */
+    public static function createWithId(int $id, int $listId = null)
+    {
+        return new static($id, null, null, $listId);
+    }
+
+    /**
+     * Constructor
```
Comment (liderman) [inline src/Model/TransactionalRequest/Receiver.php]: Need more detailed description

File: src/Model/TransactionalRequest/Receiver.php [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @return static Receiver of message
+     */
+    public static function createWithEmail(string $email, int $listId = null)
+    {
+        return new static(null, $email, null, $listId);
+    }
+
+    /**
+     * Constructor
```
Comment (liderman) [inline src/Model/TransactionalRequest/Receiver.php]: Need more detailed description

File: src/Model/TransactionalRequest/Snippet.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * Return value
+     *
+     * @return string Value
+     */
+    public function getValue(): string
+    {
+        return $this->value;
+    }
+
```
Comment (liderman) [inline src/Model/TransactionalRequest/Snippet.php]: Please remove empty line

File: src/Request/SubscribersDeleteRequest.php [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    private $id;
+
+    /**
+     * @var int|null List ID
+     */
+    private $listId;
+
+    /**
+     * Constructor
```
Comment (liderman) [inline src/Request/SubscribersDeleteRequest.php]: Need more detailed description

File: src/Request/SubscribersDeleteRequest.php [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+     *
+     * @return static Request for delete subscriber
+     */
+    public static function createFromId(int $id, int $listId = null)
+    {
+        return new static(null, $id, $listId);
+    }
+
+    /**
+     * Constructor
```
Comment (liderman) [inline src/Request/SubscribersDeleteRequest.php]: Need more detailed description

File: src/Request/SubscribersPostRequest.php [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+use Citilink\ExpertSenderApi\Exception\ExpertSenderApiException;
+use Citilink\ExpertSenderApi\Model\SubscribersRequest\Options;
+use Citilink\ExpertSenderApi\Model\SubscribersRequest\SubscriberInfo;
+use Citilink\ExpertSenderApi\RequestInterface;
+use Citilink\ExpertSenderApi\Utils;
+
+/**
+ * Request for add/edit subscriber
+ *
+ * @see https://manual.expertsender.ru/metody/podpischiki/dobavlenie-podpischika-v-list Документация
```
Comment (liderman) [inline src/Request/SubscribersPostRequest.php]: Please use English

File: src/Request/SubscribersPostRequest.php [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+use Citilink\ExpertSenderApi\Enum\SubscribersRequest\Mode;
+use Citilink\ExpertSenderApi\Exception\ExpertSenderApiException;
+use Citilink\ExpertSenderApi\Model\SubscribersRequest\Options;
+use Citilink\ExpertSenderApi\Model\SubscribersRequest\SubscriberInfo;
+use Citilink\ExpertSenderApi\RequestInterface;
+use Citilink\ExpertSenderApi\Utils;
+
+/**
+ * Request for add/edit subscriber
+ *
```
Comment (liderman) [inline src/Request/SubscribersPostRequest.php]: Doc block `@author`?

File: src/Request/SubscribersPostRequest.php [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+
+    /**
+     * Create request for edit subscriber by email
+     *
+     * @param string $email Email
+     * @param int $listId List ID
+     * @param SubscriberInfo $subscriberData Subscriber data
+     * @param Mode|null $mode Adding mode
+     * @param Options|null $options Options
+     *
```
Comment (liderman) [inline src/Request/SubscribersPostRequest.php]: Specify that the method can throw an exception

File: src/Request/TransactionalPostRequest.php [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            foreach ($this->attachments as $attachment) {
+                $xmlWriter->startElement('Attachment');
+                $xmlWriter->writeElement('FileName', $attachment->getFilename());
+                $xmlWriter->writeElement('Content', $attachment->getContent());
+                if (!empty($attachment->getMimeType())) {
+                    $xmlWriter->writeElement('MimeType', $attachment->getMimeType());
+                }
+
+                $xmlWriter->endElement(); // Attachment
+            }
```
Comment (liderman) [inline src/Request/TransactionalPostRequest.php]: Pleas add empty line after `}`

File: src/Response.php [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    private $errorCode;
+
+    /**
+     * @var string|null Error message
+     */
+    private $errorMessage;
+
+    /**
+     * Constructor
```
Comment (liderman) [inline src/Response.php]: Need a detailed description

File: src/Response/SubscribersGetShortResponse.php [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    public function isInBlackList()
+    {
+        if (!$this->isOk()) {
+            throw TryToAccessDataFromErrorResponseException::createFromResponse($this);
+        }
+
+        $xml = $this->getSimpleXml();
+        $nodes = $xml->xpath('/ApiResponse/Data/BlackList');
+        if (count($nodes) === 0) {
+            throw ParseResponseException::createFromResponse('Невозможно найти элемент BlackList', $this);
```
Comment (liderman) [inline src/Response/SubscribersGetShortResponse.php]: Please use English

File: src/Response/TableDataResponse.php [hunk 17]
```diff
@@ -1,27 +1,31 @@
...
+use Citilink\ExpertSenderApi\ResponseInterface;
+use Citilink\ExpertSenderApi\SpecificMethodResponse;
 
-class TableDataResult extends ApiResult
+class TableDataResponse extends SpecificMethodResponse
 {
-
-    /** @var array */
+    /**
+     * @var array
```
Comment (liderman) [inline src/Response/TableDataResponse.php]: Documentation?

File: src/Response/TableDataResponse.php [hunk 18]
```diff
@@ -1,27 +1,31 @@
 <?php
```
Comment (liderman) [inline src/Response/TableDataResponse.php]: `declare(strict_types=1);`?

File: src/Response/TableDataResponse.php [hunk 19]
```diff
@@ -1,27 +1,31 @@
...
 
-namespace LinguaLeo\ExpertSender\Results;
+namespace Citilink\ExpertSenderApi\Response;
 
-use LinguaLeo\ExpertSender\ApiResult;
+use Citilink\ExpertSenderApi\ResponseInterface;
+use Citilink\ExpertSenderApi\SpecificMethodResponse;
 
-class TableDataResult extends ApiResult
+class TableDataResponse extends SpecificMethodResponse
```
Comment (liderman) [inline src/Response/TableDataResponse.php]: Documentation?

File: src/Response/TableDataResponse.php [hunk 20]
```diff
@@ -1,27 +1,31 @@
...
+     * @param ResponseInterface $response Ответ от ExpertSender API
      */
-    public function __construct($response)
+    public function __construct(ResponseInterface $response)
     {
         parent::__construct($response);
         $this->parse();
     }
 
     protected function parse()
```
Comment (liderman) [inline src/Response/TableDataResponse.php]: Documentation?

File: src/Response/TableDataResponse.php [hunk 21]
```diff
@@ -1,27 +1,31 @@
...
+    /**
+     * @var array
+     */
     protected $data = [];
 
     /**
-     * @param \LinguaLeo\ExpertSender\ExpertSenderResponse $response
+     * Constructor
+     *
+     * @param ResponseInterface $response Ответ от ExpertSender API
```
Comment (liderman) [inline src/Response/TableDataResponse.php]: Please use English

File: src/Response/TimeGetResponse.php [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Time of server response
+ *
+ * @author Nikita Sapogov <sapogov.n@citilink.ru>
+ */
+class TimeGetResponse extends SpecificMethodResponse
+{
+    /**
+     * Возвращает дату сервера
```
Comment (liderman) [inline src/Response/TimeGetResponse.php]: Please use English

File: src/Response/TimeGetResponse.php [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+                if ($xmlReader->nodeType === \XMLReader::TEXT) {
+                    $dataAsString = $xmlReader->value;
+                    break;
+                }
+            }
+        }
+
+        if ($dataAsString === null) {
+            throw ParseResponseException::createFromResponse(
+                'Не удалось найти элемент Data или содержимое этого элемента',
```
Comment (liderman) [inline src/Response/TimeGetResponse.php]: Please use English

File: src/Response/TimeGetResponse.php [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+                $this
+            );
+        }
+
+        try{
+            return new \DateTime($dataAsString);
+        } catch (\Exception $e) {
+            throw new ParseResponseException(
+                sprintf('Не удалось создать объект даты из строки "%s"', $dataAsString),
```
Comment (liderman) [inline src/Response/TimeGetResponse.php]: Please use English

File: src/Response/TimeGetResponse.php [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * Time of server response
+ *
+ * @author Nikita Sapogov <sapogov.n@citilink.ru>
+ */
+class TimeGetResponse extends SpecificMethodResponse
+{
+    /**
+     * Возвращает дату сервера
+     *
```
Comment (liderman) [inline src/Response/TimeGetResponse.php]: It is necessary to specify that the method can throws an exception

File: src/Response/TimeGetResponse.php [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+     * @return \DateTime Дата сервера
```
Comment (liderman) [inline src/Response/TimeGetResponse.php]: Please use English

File: src/Response/TransactionalPostResponse.php [hunk 27]
```diff
@@ -0,0 +1,35 @@
+<?php
+
```
Comment (liderman) [inline src/Response/TransactionalPostResponse.php]: `declare(strict_types=1);`?

File: src/Response/UserIdResult.php [hunk 28]
```diff
@@ -1,11 +1,11 @@
 <?php
```
Comment (liderman) [inline src/Response/UserIdResult.php]: `declare(strict_types=1);`?

File: src/Response/UserIdResult.php [hunk 29]
```diff
@@ -1,11 +1,11 @@
...
-namespace LinguaLeo\ExpertSender\Results;
+namespace Citilink\ExpertSenderApi\Response;
 
-use LinguaLeo\ExpertSender\ApiResult;
-use LinguaLeo\ExpertSender\ExpertSenderException;
+use Citilink\ExpertSenderApi\Exception\ExpertSenderApiException;
+use Citilink\ExpertSenderApi\SpecificMethodResponse;
 
-class UserIdResult extends ApiResult
+class UserIdResult extends SpecificMethodResponse
```
Comment (liderman) [inline src/Response/UserIdResult.php]: Please add documentation

File: src/Response/UserIdResult.php [hunk 30]
```diff
@@ -1,11 +1,11 @@
...
 {
     protected $id;
```
Comment (liderman) [inline src/Response/UserIdResult.php]: Doc?

File: src/Response/UserIdResult.php [hunk 31]
```diff
@@ -18,12 +18,12 @@ public function __construct($response)
 
     public function parseBody()
```
Comment (liderman) [inline src/Response/UserIdResult.php]: Doc?

File: src/Utils.php [hunk 32]
```diff
@@ -0,0 +1,24 @@
...
+class Utils
+{
+    /**
+     * Convert bool value to string equivalent for ExpertSender API
+     *
+     * @param bool $bool Value
+     *
+     * @return string String equivalent of bool value for ExpertSender API
+     */
+    public static function convertBoolToStringEquivalent(bool $bool): string
```
Comment (liderman) [inline src/Utils.php]: native function `strval`?

File: tests/Chunk/DataChunkTest.php [hunk 33]
```diff
@@ -0,0 +1,21 @@
...
+        $bodyChunk2 = new SimpleChunk('name', 'data2');
+
+        $dataChunk = new DataChunk([$bodyChunk1, $bodyChunk2], 'subscriber');
+
+        $text = $dataChunk->toXml();
+        $this->assertRegExp('~subscriber~', $text);
+        $this->assertRegExp('~data1~', $text);
+        $this->assertRegExp('~data2~', $text);
+    }
+}
```
Comment (liderman) [inline tests/Chunk/DataChunkTest.php]: Add empty line after `}`

File: tests/Chunk/HeaderChunkTest.php [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+    public function testGetText()
+    {
+        $bodyChunk = new SimpleChunk('name', 'body');
+        $headerChunk = new HeaderChunk('api-key', [$bodyChunk]);
+        $text = $headerChunk->toXml();
+
+        $this->assertRegExp('~body~', $text);
+        $this->assertRegExp('~api-key~', $text);
+    }
+}
```
Comment (liderman) [inline tests/Chunk/HeaderChunkTest.php]: Empty line

File: src/Chunk/SimpleChunk.php [hunk 35]
```diff
@@ -0,0 +1,59 @@
...
+     * Constructor
+     *
+     * @param string $name Name
+     * @param bool $value Value
+     *
+     * @return static Simple chunk
+     */
+    public static function createBool(string $name, bool $value)
+    {
+        return new static($name, $value ? 'true' : 'false');
```
Comment (AmsTaFFix) [inline src/Chunk/SimpleChunk.php]: `strval(true) === '1'` vs `(true ? 'true' : 'false') === 'true'`

File: src/Response/TableDataResponse.php (hunk 18 above)
Comment (AmsTaFFix) [inline src/Response/TableDataResponse.php]: it's old code, it can't work with strict types

File: src/Response/TableDataResponse.php (hunk 17 above)
Comment (AmsTaFFix) [inline src/Response/TableDataResponse.php]: it's old code, it will be deleted soon

File: src/Response/TableDataResponse.php (hunk 20 above)
Comment (AmsTaFFix) [inline src/Response/TableDataResponse.php]: it's old code, it will be deleted soon

File: src/Response/UserIdResult.php (hunk 30 above)
Comment (AmsTaFFix) [inline src/Response/UserIdResult.php]: it's old code and will be deleted soon

File: src/Response/UserIdResult.php (hunk 31 above)
Comment (AmsTaFFix) [inline src/Response/UserIdResult.php]: it's old code and will be deleted soon

File: src/Utils.php (hunk 32 above)
Comment (AmsTaFFix) [inline src/Utils.php]: i already wrote about it earlier

File: src/Response/UserIdResult.php (hunk 28 above)
Comment (AmsTaFFix) [inline src/Response/UserIdResult.php]: it's old code, it will be deleted soon

File: tests/Chunk/DataChunkTest.php (hunk 33 above)
Comment (AmsTaFFix) [inline tests/Chunk/DataChunkTest.php]: already did it

File: tests/Chunk/HeaderChunkTest.php (hunk 34 above)
Comment (AmsTaFFix) [inline tests/Chunk/HeaderChunkTest.php]: already did it

File: src/Enum/SubscribersRequest/MatchingMode.php [hunk 36]
```diff
@@ -0,0 +1,37 @@
+<?php
+declare(strict_types=1);
+
+namespace Citilink\ExpertSenderApi\Enum\SubscribersRequest;
+
+use MyCLabs\Enum\Enum;
+
+/**
+ * Matching mode in add/edit subscriber requests
+ *
```
Comment (AmsTaFFix) [inline src/Enum/SubscribersRequest/MatchingMode.php]: already did it

File: src/Request/SubscribersPostRequest.php (hunk 12 above)
Comment (AmsTaFFix) [inline src/Request/SubscribersPostRequest.php]: removed russian language from SubscribersGetShortResponse

File: src/Request/SubscribersPostRequest.php (hunk 13 above)
Comment (AmsTaFFix) [inline src/Request/SubscribersPostRequest.php]: already did it

File: src/Request/TransactionalPostRequest.php (hunk 14 above)
Comment (AmsTaFFix) [inline src/Request/TransactionalPostRequest.php]: already did it

**Pull request merged**